mailpoet/tests/unit/Subscribers/SubscriberSubscribeControllerUnitTest.php

```diff
@@ -0,0 +1,494 @@
+<?php
+declare(strict_types=1);
+
+namespace MailPoet\Subscribers;
+
+use Codeception\Stub;
+use Codeception\Stub\Expected;
+use MailPoet\Entities\FormEntity;
+use MailPoet\Entities\SubscriberEntity;
+use MailPoet\Form\FormsRepository;
+use MailPoet\Form\Util\FieldNameObfuscator;
+use MailPoet\Settings\SettingsController;
+use MailPoet\Statistics\StatisticsFormsRepository;
+use MailPoet\Subscription\Captcha;
+use MailPoet\Subscription\CaptchaSession;
+use MailPoet\Subscription\SubscriptionUrlFactory;
+use MailPoet\Subscription\Throttling as SubscriptionThrottling;
+use MailPoet\UnexpectedValueException;
+use MailPoet\WP\Functions as WPFunctions;
+
+class SubscriberSubscribeControllerUnitTest extends \MailPoetUnitTest {
+  public function testErrorGetsThrownWhenEmailFieldIsNotObfuscated() {
+    $subscriptionCaptcha = Stub::makeEmpty(Captcha::class);
+    $captchaSession = Stub::makeEmpty(CaptchaSession::class);
+    $subscriberActions = Stub::makeEmpty(
+      SubscriberActions::class,
+      [
+        'subscribe' => Expected::never(),
+      ]
+    );
+    $subscriptionUrlFactory = Stub::makeEmpty(SubscriptionUrlFactory::class);
+    $throttling = Stub::makeEmpty(SubscriptionThrottling::class);
+    $fieldNameObfuscator = Stub::makeEmpty(FieldNameObfuscator::class);
+    $requiredCustomFieldValidator = Stub::makeEmpty(RequiredCustomFieldValidator::class);
+    $settings = Stub::makeEmpty(SettingsController::class);
+    $form = Stub::makeEmpty(FormEntity::class);
+
+    $formsRepository = Stub::makeEmpty(
+      FormsRepository::class,
+      [
+        'findOneById' => function() use ($form): FormEntity {
+          return $form;
+        },
+      ]
+    );
+    $statisticsFormsRepository = Stub::makeEmpty(
+      StatisticsFormsRepository::class,
+      [
+        'subscribe' => Expected::never(),
+      ]
+    );
+
+    $wp = Stub::make(
+      WPFunctions::class,
+      [
+        'doAction' => Expected::never(),
+      ],
+      $this
```

Comment:
Just a tip. I see you are using the third `$this` argument in stubs intermittently: sometimes it's here, sometimes it's not. There's a gotcha that expectations may not get executed if `$this` argument is omitted, it's not immediately obvious with `Expected::never()` because if it doesn't work it's kind of expected, but with `Expected::once()` etc. you may see the problem right away. So I would pay attention to that or even maybe build a habit of always adding this argument if there are any expectations, just to be safe.

Reply:
Addressed in 11d934d